libraries/abstractions/platform/freertos/iot_network_udp.c

```diff
@@ -0,0 +1,534 @@
+    {
+        IotLogError( " Failed to set UDP socket timeout status = %d ", socketStatus );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_NETWORK_FAILURE );
+    }
+
+    udpServerAddr = SOCKETS_GetHostByName( pServerInfo->pHostName );
+
+    if( udpServerAddr == 0 )
+    {
+        IotLogError( "Failed to resolve DNS address %s.",
+                     pServerInfo->pHostName );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_NETWORK_FAILURE );
+    }
+
+    *pNetworkConnection = pvPortMalloc( sizeof( _networkConnection_t ) );
+
+    if( *pNetworkConnection == NULL )
+    {
+        IotLogError( "Failed to allocate memory for new network connection." );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_NETWORK_NO_MEMORY );
+    }
+
+    /* Clear the connection information. */
+    ( void ) memset( *pNetworkConnection, 0x00, sizeof( _networkConnection_t ) );
+
+    ( *pNetworkConnection )->socket = udpSocket;
+    ( *pNetworkConnection )->serverAddr.ulAddress = udpServerAddr;
+    ( *pNetworkConnection )->serverAddr.usPort = SOCKETS_htons( pServerInfo->port );
+    ( *pNetworkConnection )->serverAddr.ucSocketDomain = SOCKETS_AF_INET;
+    ( *pNetworkConnection )->strictServerAddrCheck = pServerInfo->strictServerAddrCheck;
+
+    /* Create the connection event flags and mutex. */
+    pConnectionFlags = xEventGroupCreateStatic( &( ( *pNetworkConnection )->connectionFlags ) );
+
+    /* Static event flags and mutex creation should never fail. The handles
+     * should point inside the connection object. */
+    configASSERT( pConnectionFlags == ( EventGroupHandle_t ) &( ( *pNetworkConnection )->connectionFlags ) );
+
+
+    IOT_FUNCTION_CLEANUP_BEGIN();
+
+    if( status != IOT_NETWORK_SUCCESS )
+    {
+        if( udpSocket != SOCKETS_INVALID_SOCKET )
+        {
+            SOCKETS_Close( udpSocket );
+        }
+
+        /* Clear the connection information. */
+        if( (pNetworkConnection != NULL) && ( ( *pNetworkConnection ) != NULL ))
+        {
+            vPortFree( *pNetworkConnection );
+        }
+    }
+
+    IOT_FUNCTION_CLEANUP_END();
+}
+/*-----------------------------------------------------------*/
+
+IotNetworkError_t IotNetworkUdp_SetReceiveCallback( void * pConnection,
+                                                    IotNetworkReceiveCallback_t receiveCallback,
+                                                    void * pContext )
+{
+    IotNetworkError_t status = IOT_NETWORK_SUCCESS;
+
+    /* Cast network connection to the correct type. */
+    _networkConnection_t * pNetworkConnection = ( _networkConnection_t * ) pConnection;
+
+    /* Set the receive callback and context. */
+    pNetworkConnection->receiveCallback = receiveCallback;
+    pNetworkConnection->pReceiveContext = pContext;
+
+    /* No flags should be set. */
+    configASSERT( xEventGroupGetBits( ( EventGroupHandle_t ) &( pNetworkConnection->connectionFlags ) ) == 0 );
+
+    /* Create task that waits for incoming data. */
+    if( xTaskCreate( _networkReceiveTask,
```

Comment:
what if ```IotNetworkUdp_SetReceiveCallback``` is called twice? Would it end up creating 2 tasks?

Reply:
But that means, that ```receiveCallback``` and ```pReceiveContext``` from 1st application will be overwritten by the call from 2nd application. How does this work when 2 applications are using this interface?

Reply:
Please see my comments below regarding creating new tasks - in my opinion no tasks should be created at all if the FreeRTOS+TCP stack is used - can't comment for lwIP though.

Reply:
> Please see my comments below regarding creating new tasks - in my opinion no tasks should be created at all if the FreeRTOS+TCP stack is used - can't comment for lwIP though.

Network abstraction for TCP currently creates task for asynchronous receive if callback is set.  UDP abstraction code was written to be consistent with TCP network abstraction code:  https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/platform/freertos/iot_network_freertos.c#L458
Network abstraction is common for all TCP/IP stacks, it uses AFR secure sockets interface.

Reply:
> But that means, that `receiveCallback` and `pReceiveContext` from 1st application will be overwritten by the call from 2nd application. How does this work when 2 applications are using this interface?

Each application will use different pConnection. IotNetworkUdp_Create() allocates new connection, so there no overwriting. If same application calls it twice (with same pConnection), it will be overwritten and that is intentional.

Reply:
Each ( _networkConnection_t * ) pConnection has a delicate socket and send/recv task. Additional call for this function with different pConnection will be on different socket and have new recv task.